feat: Autobrr

ludviglundgren opened the pull request:
## Description
Add support for `autobrr`, which is a modern alternative to autodl-irssi with built in support for new clients and more coming.
It's very much still work in progress and can't be considered stable yet. 
https://github.com/autobrr/autobrr

!! Pulled latest into the branch so unfortunately more stuff in it than only my things.

## Fixes issues: 
- Un-tracked issue: Add autobrr

## Proposed Changes:
- Add autobrr installer, remover and updater

## Change Categories
- New feature

## Checklist
<!-- Please note that we also require you to check the CONTRIBUTORS.md file, this is just a short list-->
- [x] Branch was made off the `develop` branch and the PR is targetting the `develop` branch
- [x] Docs have been made OR are not necessary
    - PR link: https://github.com/liaralabs/docs.swizzin.ltd/pull/98
- [x] Changes to panel have been made OR are not necessary
    - PR link: https://github.com/swizzin/swizzin_dashboard/pull/39
- [x] Code is formatted [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Code conforms to project structure [(See more)](https://swizzin.ltd/dev/structure)
- [x] Shellcheck isn't screaming [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Prints to terminal are handled [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#printing-into-the-terminal)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Testing was done
   - [x] Tests created or no new tests necessary
   - [x] Tests executed

## Test scenarios
<!-- Please let us know what has been done or anything else that works/doesn't. Feel free to copy-paste the examples at the bottom of this section -->
Did manual testing only remote vm and local vm.

### Architectures
<!--
Please use these emojis here to fill the table below. It will nicely auto-format with spacing, don't worry. Leave empty wherever you do not know / have not tested
✅ = Works successfully
❎ = Does not work BUT is handled gracefully
🛠 = Still WIP
❌ = Broken / not working
-->
It has multi arch support thanks to golang but have not yet tested and verified other than amd64.

|   			| `amd64` 	| `armhf` 	| `arm64` 	| Unspecified 	|
|--------		|-------- 	|-------- 	|-------- 	|----------		|
| Focal 		|			|			|			|				|
| Bionic		|			|			|			|				|
| Buster		|	✅ 		|			|	 		|				|
| Stretch		|			|			|			|				|
| Raspbian  	|	⚫️		|			|	⚫️		|	⚫️			|

### ✅❎ Passed

### 🛠🛠 TODO

- [x] Multiuser support

### ❌❌ Currently failing

<!-- EXAMPLES :
- Fresh app install without nginx
    - With only one user
    - With multiple users present
- Fresh Install with nginx present nginx
    - With only one user
    - With multiple users present
- Fresh install and nginx install afterwards
    - With only one user
    - With multiple users present
- Update from version in master
- Upgrade from version in master
- Password gets changed from `box` in app
- User removal from `box` acting on app
- New user in `box` gets added to app
- Sysinfo compatibility
    - Info washed
    - Content available
-->


// ze0s

flying-sausages commented:
Heya, thanks for the PR, looks like a good fit for our project. If you do `git rebase origin/develop` and then `git push --force` you should get rid of the dupe commits.

I can also do that for you if you allowed maintainers to push to this branch. Let me know, and I can start reviewing.

How stable is the current build of `autobrr`? Do you have a roadmap of some sorts? FYI, We tend to delay merging applications which are in alpha/nightly stage (e.g. #712) and wait for a beta/development level of maturity; mostly because of support.

ludviglundgren commented:
Hello!

I added you as a maintainer so you should be able to do the cleanup. 

The latest release, v0.1.0 is stable enough to use, but more features are incoming and there might be breaking changes before a v1 is released. I'll try not to do it, but could happen. 

No roadmap yet, but some of the stuff on the todo list:

* add support for more trackers
* add support for Deluge
* add support for running commands on filtered (like autodl-irssi)
* add support to send to sonarr/radarr etc
* more advanced scheduling, rules like x per hour, max 2 downloads at a time and so on
* documentation! And this will probably keep people away for a while, but would be nice to have the installer in swizzin for easy deploy or releases.
* notifications
* automatic management like moving files, delete on X condition
* filter options specific to music, games and maybe more

brettpetch commented:
- [ ] Make Testing Suite (should be able to with the code i supplied in WS if it's MU)
- [ ] Multiuser
- [ ] Some kind of app auth?

ludviglundgren commented:
Fixed perms, and ran the suggested formatters.

ludviglundgren commented:
After a discussion with B we came to the conclusion that adding app level auth and making it per user/swizzin multuser will be the way to go. I'll start working on it and notify when it's done.

ludviglundgren commented:
Added 

> * [ ]  Make Testing Suite (should be able to with the code i supplied in WS if it's MU, if not you could just use the swizzdb var to get the owner user)
> * [ ]  Multiuser
> * [ ]  Some kind of app auth?
> * [ ]  Fix script perms (644 currently, should be 755)
> 
> Nice work otherwise though!

Fixed:

 * [x]  Multiuser ✅ 
 * [x]  Some kind of app auth? Built it's own auth system ✅ 
 * [x]  Fix script perms (644 currently, should be 755) ✅ 

Still unsure about the "Make testing suite". 
Still no docs, so if we can with with them that would be great.

flying-sausages commented:
> Still unsure about the "Make testing suite".

`box test` could probably guess most of what it needs to be default, try installing it and then run the command and then we can see what "exemptions" need to be made for it

brettpetch commented:
It’s an MU app, wouldn’t it need to be manual?

/b

> 
> On Aug 15, 2021, at 4:21 PM, flying-sausages ***@***.***> wrote:
> 
> ﻿
> Still unsure about the "Make testing suite".
> 
> box test could probably guess most of what it needs to be default, try installing it and then run the command and then we can see what "exemptions" need to be made for it
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

flying-sausages commented:
Depends on where the auth is handled, i.e. whether it's one process running or one process per user

brettpetch commented:
> Depends on where the auth is handled, i.e. whether it's one process running or one process per user

It’s thru nginx now ;)

ludviglundgren commented:
Multiuser with per app install and auth per user. Some of the box test failed for it but not the nginx and curl test.

flying-sausages commented:
> Multiuser with per app install and auth per user. Some of the box test failed for it but not the nginx and curl test.

Cool, I can take a look once the deployment method is finalised

ludviglundgren commented:
@brettpetch helped me with the setup and if he doesn't have anything to add then it's done from my side.

flying-sausages commented:
> @brettpetch helped me with the setup and if he doesn't have anything to add then it's done from my side.

oh shit I might have just possibly accidentally pushed over some of your commits while i was rebasing on top of develop

flying-sausages commented:
>  Still no docs, so if we can with with them that would be great.

This is another docs app addition PR from earlier https://github.com/liaralabs/docs.swizzin.ltd/pull/81/files

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+#shellcheck source=sources/functions/utils
+. /etc/swizzin/sources/functions/utils
+
+app_reqs=("curl" "sqlite3")
+apt_install "${app_reqs[@]}"
```
I see where you pulled this from, but that's an "arr" format, tl;dr:
```suggestion
apt_install curl sqlite3
```

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+
+users=($(_get_user_list))
+
+if [[ -n $1 ]]; then
+    user=$1
+    _autobrr_user_config ${user}
+    if [[ -f /install/.nginx.lock ]]; then
+        echo_progress_start "Configuring nginx"
+        bash /etc/swizzin/scripts/nginx/autobrr.sh
+        systemctl reload nginx
+        echo_progress_done
+    fi
+    exit 0
+fi
```
This whole section might have to go under the function definitions as `_autobrr_user_config` will not be defined at this point

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+    echo_progress_start "Extracting archive"
+
+    # the archive contains both autobrr and autobrrctl to easily setup the user
+    tar xfv "/tmp/autobrr.tar.gz" --directory /usr/bin/ >> "$log" 2>&1 || {
```
no need to ensure perms to `+x`? if the perms are stored in the archive then it won't be an issue, just curious

ludviglundgren replied:
The packager handles all that so the binaries will always be executable.

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+    sessionSecret="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c16)"
+
+    if [ ! -d "/home/$user/.config/autobrr/" ]; then
+        mkdir -p "/home/$user/.config/autobrr/"
```
Just in case this is the first app a user installs (it starting with `a`, high chance), gotta make sure that `~/.config` is not owned by `root`
```suggestion
        mkdir -p "/home/$user/.config/autobrr/"
        chown "$user": /home/"$user"/.config
```

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+    fi
+    chown -R "$user": "/home/$user/.config/autobrr"
+
+    cat > "/home/$user/.config/autobrr/config.toml" << CFG
```
Not a requirement or anything but if the `#` lines are not necessary, I'd just wipe them out and add a link to the full content of the default or some wiki page where users can find more.

ludviglundgren replied:
I'd like to keep it in the config.

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+    if [ ! -d "/home/$user/.config/autobrr/" ]; then
+        mkdir -p "/home/$user/.config/autobrr/"
+    fi
+    chown -R "$user": "/home/$user/.config/autobrr"
```
L113 dupe, we only need the last one of these that will run

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+
+    for user in "${users[@]}"; do
+        echo_log_only "Adding user $user"
+        pass=$(_get_user_password "$user")
+
+        # the password needs to be created with argon2 so we use autobrrctl to create the user
+        # using sqlite3 directly was not an option
+        echo -n "$pass" | autobrrctl --config "/home/$user/.config/autobrr" create-user "$user"
```
pipe pipe pipe
```suggestion
        # the password needs to be created with argon2 so we use autobrrctl to create the user
        # using sqlite3 directly was not an option
        _get_user_password "$user" | autobrrctl --config "/home/$user/.config/autobrr" create-user "$user"
```

ludviglundgren replied:
Great improvement!

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+}
+
+_systemd_autobrr() {
+    if [[ ! -f /etc/systemd/system/autobrr@.service ]]; then
```
I'd just _always_ overwrite the systemd service if this is being installed. Otherwise some edge case could keep on re-using an out of date service file. i.e. I'd remove this `if fi`

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+User=%i
+Group=%i
+ExecStart=/usr/bin/autobrr --config=/home/%i/.config/autobrr/
+# TimeoutStopSec=20
+# KillMode=process
+# Restart=on-failure
```
Are these still to decide? Otherwise I'd yank 'em out or use them

ludviglundgren replied:
Yeah I'm no super familiar with all systemd options, so this was copied from another service. I have removed them now since I'm not sure if they are needed or not.

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+        bash /etc/swizzin/scripts/nginx/autobrr.sh
+        systemctl reload nginx
+        echo_progress_done
+    fi
```
The config function does not start the service it seems
```suggestion
    fi
    systemctl enable -q --now autobrr@${user} 2>&1 | tee -a $log
```

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+
+_install_autobrr
+
+_systemd_autobrr
+_systemd_autobrr_enable_for_user
```
Might as well merge these into one

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+
+_systemd_autobrr
+_systemd_autobrr_enable_for_user
+_autobrr_add_user
```
I'd add an `s` to the end to make this a bit more different to `_autobr_config_user` (seeing as it also processes more than one user)
```suggestion
_autobrr_add_users
```

flying-sausages replied:
(this will also need the actual function def name to be changed, mind you)

ludviglundgren replied:
Renamed and done.

flying-sausages commented on scripts/nginx/autobrr.sh:
```diff
@@ -0,0 +1,36 @@
+if [[ ! -f /etc/nginx/apps/autobrr.conf ]]; then
+
+    cat > /etc/nginx/apps/autobrr.conf << 'AUTOBRR'
+location /brr {
```
lol :p are you sure `/brr` 301ing to `/autobrr/` won't be a problem with double-slashes? I never know myself about this when it comes to these nginx things

ludviglundgren replied:
Maybe not. Removed it 👍

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+#
+# Default: "localhost"
+#
+host = "127.0.0.1"
```
Is it ok for this to be on `127.0.0.1` if someone does not use `nginx`? We usually allow this to be `*` or similar by default, and then in the event of `nginx` we change it to `127...1` only. We _did_ see some issues with this recently in the arrs, @liaralabs any advice on this? Some call to make to also apply going forward?

ludviglundgren replied:
The thought before was to put it behind basic auth on the reverse proxy, but now that the app have auth we could potentially expose it on 0.0.0.0.

flying-sausages replied:
Yeah, I think that would be optimal. As long as it gets changed to `127.0.0.7` from the script for `nginx`

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+#
+# Optional
+#
+#logPath = "log/autobrr.log"
```
This would be nice to enable, `~/.config/autobrr/log.txt` or similar would be good enough. Especially if the output is set to `DEBUG`, that could fill up the `journalctl` fast. Had that happen with transmission before (spamming stdout while in service) and it just _kills_ performance of all `systemctl` etc. commands

ludviglundgren replied:
Interesting! Currently it outputs to both the console and logfile if set. It has mechanisms to rotate logs as well so it doesn't fill up the disk. Will have to look into writing only to log file if set, and nothing to console then.

ludviglundgren replied:
I set it to log to log to `/home/$user/.config/autobrr/logs/autobrr.log` and will make some changes to only have some stuff output to console/stdout. But that will come with an update to the app :)

flying-sausages commented on scripts/upgrade/autobrr.sh:
```diff
@@ -0,0 +1,60 @@
+    exit 1
+fi
+
+_upgrade_autobrr() {
```
I noticed this code is duplicate from the install script, this could be de-duped by defining the function once in `sources/functions/autobrr` and then using it both in the install script as well as here

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,189 @@
+
+users=($(_get_user_list))
+
+if [[ -n $1 ]]; then
```
You will need to modify the useradd function inside `script/box` in order for it to trigger this. Also, please make a case for the password changing mechanism in the same file, so that the passwords are kept in sync everywhere when they're changed

ludviglundgren replied:
Would that be added here? https://github.com/swizzin/swizzin/blob/6f07eb7f1ed54ccea217d01d45c658b1fbc8b1be/scripts/box#L306-L331

Re the password changing mech, I'll add a function for that to `autobrrctl`.

flying-sausages replied:
Yup, that's the one for `box adduser`. `box chpasswd` and `box deluser` work in a similar-ish fashion

ludviglundgren replied:
It only seems to be needed to add in the list at 306 right? Because the line 317 would trigger it, so extra check needed?

flying-sausages replied:
Yup! If it has no special procedure, all it needs is to be in the list

flying-sausages reviewed (changes requested):
That's a good-looking PR! First read-through, these are some things that popped out at me. Gonna give this a run on a fresh machine tomorrow.

flying-sausages commented on scripts/nginx/autobrr.sh:
```diff
@@ -0,0 +1,36 @@
+. /etc/swizzin/sources/functions/utils
+users=($(_get_user_list))
+
+if [[ ! -f /etc/nginx/apps/autobrr.conf ]]; then
```
I would also not check the conditional here, just write it whether it exists or not

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,146 @@
+
+        # the password needs to be created with argon2 so we use autobrrctl to create the user
+        # using sqlite3 directly was not an option
+        _get_user_password "$user" | autobrrctl --config "/home/$user/.config/autobrr" create-user "$user"
```
I think `autobrrctl` needs to have the full path specified here btw, it's not showing up in my `path`

<img width="696" alt="afbeelding" src="https://user-images.githubusercontent.com/23618693/130102913-4d57dedf-029f-4785-82bc-fae5d26575a7.png">

flying-sausages commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,146 @@
+    fi
+}
+
+autobrr_download_latest
```
The function still needs to be pulled in
```suggestion
#shellcheck source=sources/functions/autobrr
. /etc/swizzin/sources/functions/autobrr
autobrr_download_latest
```

flying-sausages commented on scripts/upgrade/autobrr.sh:
```diff
@@ -0,0 +1,24 @@
+    echo_progress_done "Service restarted"
+}
+
+autobrr_download_latest
```
Function missing here
```suggestion
#shellcheck source=sources/functions/autobrr
. /etc/swizzin/sources/functions/autobrr
autobrr_download_latest
```

flying-sausages commented:
@brettpetch @liaralabs can one of you please do another full check that this performs as it should? I would say this is good to merge to develop

liaralabs commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        # Only run the nginx hook if a new user is being added, otherwise it will be ran for all users at the end of the installer
+        if [[ ${#users[@]} -eq 1 ]]; then
+            if [[ -f /install/.nginx.lock ]]; then
+                bash /etc/swizzin/scripts/nginx/autobrr.sh
+                systemctl reload nginx
+            fi
+        fi
```
This snippet will fire if the master user is the only user on the seedbox when autobrr is installed.

It's a bit more clunky, but a variable could be set between lines 120-124 that indicates that `adduser` is the current environment rather than `install`. That is, assuming that double nginx setup is an issue; however, in theory it should not be

flying-sausages replied:
resolved, logic moved into the main's `if $1` statement so it's outside the loop

liaralabs commented on scripts/install/autobrr.sh:
```diff
@@ -0,0 +1,134 @@
+#
+sessionSecret = "${sessionSecret}"
+CFG
+        chown -R "$user": "/home/$user/.config/autobrr"
```
Run this command after creating the user (line 98) or suffer read-only root owned db :smile:
